### PR TITLE
Exception group stage1 (by @iritkatriel)

### PR DESCRIPTION
**Pull request by gvanrossum**
This PR is just so I can more easily comment...

### CODE REVIEW

File: Lib/types.py [hunk 1]
```diff
@@ -300,4 +300,81 @@ def wrapped(*args, **kwargs):
 NoneType = type(None)
 NotImplementedType = type(NotImplemented)
 
+class TracebackGroup:
```
Comment (gvanrossum) [inline Lib/types.py]: All this new code (TbGroup and ExGroup) should probably live in a separate file, don't you think? types.py should not have actual class definitions (it must be importable blindingly fast).

File: Lib/types.py [hunk 2]
```diff
@@ -300,4 +300,81 @@ def wrapped(*args, **kwargs):
...
+        # self.__traceback__ is updated as usual, but self.__traceback_group__
+        # is the frame where the exception group was created (and it is
+        # preserved on splits).  So __traceback_group__ + __traceback__
+        # gives us the full path.
+        import types
+        self.__traceback__ = types.TracebackType(None, self.frame, 0, 0)
+        self.__traceback_group__ = TracebackGroup(self.excs, self.frame)
+
+    def split(self, E):
+        ''' returns two new ExceptionGroups: match, rest
```
Comment (gvanrossum) [inline Lib/types.py]: I might as well bring up now that PEP 8 recommends a different style for docstrings (one line summary, blank line, then more text as needed -- or just a one line summary; and always using `"""`).

File: Lib/asyncio/taskgroup.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
+#
+# This source file is part of the EdgeDB open source project.
```
Comment (gvanrossum) [inline Lib/asyncio/taskgroup.py]: Hmm... @1st1, does all this need to stay in the file in case we copy it into the stdlib?

File: Lib/types.py [hunk 4]
```diff
@@ -300,4 +300,81 @@ def wrapped(*args, **kwargs):
...
+
+    def __init__(self, excs, frame=None):
+        self.excs = set(excs)
+        self.frame = frame or sys._getframe()
+        # self.__traceback__ is updated as usual, but self.__traceback_group__
+        # is the frame where the exception group was created (and it is
+        # preserved on splits).  So __traceback_group__ + __traceback__
+        # gives us the full path.
+        import types
+        self.__traceback__ = types.TracebackType(None, self.frame, 0, 0)
```
Comment (gvanrossum) [inline Lib/types.py]: Can't __traceback__ stay None? The interpreter will push a traceback item onto it when it is raised anyway.

File: Lib/types.py [hunk 5]
```diff
@@ -300,4 +300,81 @@ def wrapped(*args, **kwargs):
...
+        self.tb_frame = frame
+        self.tb_next_map = {} # TODO: make it a weak key dict
+        for e in excs:
+            self.tb_next_map[e] = e.__traceback__
+
+class ExceptionGroup(BaseException):
+
+    def __init__(self, excs, frame=None):
+        self.excs = set(excs)
+        self.frame = frame or sys._getframe()
```
Comment (gvanrossum) [inline Lib/types.py]: A shame that the exception must keep the frame alive -- normally that is done through the traceback, so at least you can sever the link by setting `__traceback__ = None`.

(In case you wonder why I care, there is a painful bit of ancient history around cycles involving exceptions and frames dating back to the Python 1.x days. This is also why `except E as e` deletes `e` at the end of the handler block.)

File: Lib/types.py [hunk 6]
```diff
@@ -300,4 +300,81 @@ def wrapped(*args, **kwargs):
...
+        self.__traceback__ = types.TracebackType(
+            self.__traceback__, frame, 0, 0)
+
+    @staticmethod
+    def render(exc, tb=None, indent=0):
+        print(exc)
+        try:
+            tb = tb or exc.__traceback__
+        except Exception as e:
+            import pdb; pdb.set_trace()
```
Comment (gvanrossum) [inline Lib/types.py]: Use `breakpoint()` (PEP 553, since Python 3.7).

File: Lib/types.py [hunk 7]
```diff
@@ -300,4 +300,81 @@ def wrapped(*args, **kwargs):
...
+                    for e, t in tbg.tb_next_map.items():
+                        print('---------------------------------------')
+                        ExceptionGroup.render(e, t, indent)
+                tb = None
+
+    def __str__(self):
+        return f"ExceptionGroup({self.excs})"
+
+    def __repr__(self):
+        return str(self)
```
Comment (gvanrossum) [inline Lib/types.py]: If you define just `__repr__`, the other will default to that.

File: Lib/types.py [hunk 8]
```diff
@@ -300,4 +300,81 @@ def wrapped(*args, **kwargs):
...
+        self.frame = frame or sys._getframe()
+        # self.__traceback__ is updated as usual, but self.__traceback_group__
+        # is the frame where the exception group was created (and it is
+        # preserved on splits).  So __traceback_group__ + __traceback__
+        # gives us the full path.
+        import types
+        self.__traceback__ = types.TracebackType(None, self.frame, 0, 0)
+        self.__traceback_group__ = TracebackGroup(self.excs, self.frame)
+
+    def split(self, E):
```
Comment (gvanrossum) [inline Lib/types.py]: Now that we have `split()`, what do you think `join()` will look like? A class method? I was also thinking perhaps it can be spelled as `+`.

**Review by gvanrossum — COMMENTED** [review]
Also for @1st1 to review.

File: Lib/types.py (hunk 8 above)
Comment (iritkatriel) [reply Lib/types.py]: Do we need join? Or do we just make a new list of exceptions (some of which can be EGs) and construct a new EG (which takes care or creating the new TBG)?

Comment (gvanrossum) [reply Lib/types.py]: Not sure, but if you split and then join, the pieces the result should be the original EG no matter how the split turned out.

File: Lib/types.py (hunk 5 above)
Comment (iritkatriel) [reply Lib/types.py]: I've removed it, we didn't really need it but I was confused.

File: Lib/types.py (hunk 4 above)
Comment (iritkatriel) [reply Lib/types.py]: Yeah this was also wrong and it's gone.

File: Lib/types.py (hunk 1 above)
Comment (iritkatriel) [reply Lib/types.py]: And also - do we want to keep the names ExceptionGroup and TracebackGroup?

Comment (gvanrossum) [reply Lib/types.py]: Let’s keep these for now.

PS. I’ll be AFK the rest of the day and much of the rest of the week.

Comment (iritkatriel) [reply Lib/types.py]: OK thanks. We'll try to make progress. Unit tests and the like.

Comment (gvanrossum) [reply Lib/types.py]: Try to make a context manager to handle exceptions, to exercise the join.

File: Lib/asyncio/taskgroup.py (hunk 3 above)
Comment (1st1) [reply Lib/asyncio/taskgroup.py]: No, I'll contribute it in a clean way under PSFL.

Comment (1st1) [reply Lib/asyncio/taskgroup.py]: Also, FWIW, the current code still doesn't work with EGs properly. If an EG contains a CancelledError, the TaskGroup would propagate the CancelledError, not the EG (thus discarding all other errors).  To fix this we'll need to fix the asyncio itself to recognize CancelledErrors wrapped in a EG. I'll do that as soon as we have some basic high-level API.

File: output.txt [hunk 9]
```diff
@@ -0,0 +1,112 @@
+Running Release|Win32 interpreter...
```
Comment (1st1) [inline output.txt]: nit: Do we need this file? They typically go quickly out of sync with the code.

**Review by 1st1 — COMMENTED** [review]
This is a good start I think.

File: output.txt (hunk 9 above)
Comment (iritkatriel) [reply output.txt]: Sure, I'll remove it and the test script as soon as we have unit tests. Which should be.. soon

**Comment by iritkatriel** [comment]
The tests are a bit of a mess still, but their coverage is decent now.

**Comment by gvanrossum** [comment]
Just curious -- did you add a merge operation yet?

**Comment by iritkatriel** [comment]
> Just curious -- did you add a merge operation yet?

There is a merge in the catcher and the tests check that each exception has the right tb but it doesn’t yet recursively merge nested EGs correctly to their previous structure.  I wasn’t sure how I would know that two nested EG are related (they were the same one before the split). Compare the frame objects by ID? If that doesn’t work we might need an EG to know the ID of the EG it was split from or the first EG it was before all the splits.

TL;DR still working on that.

**Comment by iritkatriel** [comment]
Another option is that the merge operation is specific for the catch use case, where the original (pre-split) EG is available and we use it.

**Comment by iritkatriel** [comment]
> Another option is that the merge operation is specific for the catch use case, where the original (pre-split) EG is available and we use it.

I did it the other way around - instead of merge, I added an ExceptionGroup.subgroup(exc_list) which projects an exception group to a subset of the exceptions in it. It's similar to split but divides on actual exceptions rather than types.

So in catch we decide which old exceptions we want to keep, project the EG on them, and then add the new exceptions.

**Comment by gvanrossum** [comment]
> > Another option is that the merge operation is specific for the catch use case, where the original (pre-split) EG is available and we use it.
> 
> I did it the other way around - instead of merge, I added an ExceptionGroup.subgroup(exc_list) which projects an exception group to a subset of the exceptions in it. It's similar to split but divides on actual exceptions rather than types.
> 
> So in catch we decide which old exceptions we want to keep, project the EG on them, and then add the new exceptions.

Oh, cool way of doing it. Then again,

> > Just curious -- did you add a merge operation yet?
> 
> There is a merge in the catcher and the tests check that each exception has the right tb but it doesn’t yet recursively merge nested EGs correctly to their previous structure. I wasn’t sure how I would know that two nested EG are related (they were the same one before the split). Compare the frame objects by ID? If that doesn’t work we might need an EG to know the ID of the EG it was split from or the first EG it was before all the splits.

Maybe compare the traceback objects by ID? After all they receive the same tb in project().

Also, just saw the TODOs in TracebackGroup... The solution *may* have to be to use two parallel lists, one for the keys (exceptions) and one for the values (tracebacks). And a dict mapping the exception IDs to values. And make exceptions weak refable, and in the hook remove the exception from the list and its ID from the dict. Ugh, that's pretty nasty. Maybe we need a kind of dict that is not just a weak key dict, but a weak key ID dict -- it holds the objects but compares by ID. (Hmm... That sounds so useful that I wonder if it already exists!)

**Comment by iritkatriel** [comment]
> Maybe compare the traceback objects by ID? After all they receive the same tb in project().

The tracebacks are on the EGs, so that would require the user's handler to preserve the EG structure. With what we have now the user can iterate over the individual exceptions in the EG, put some of them in a list and say "reraise these", and they will be "merged" back in with the correct structure. 

> Also, just saw the TODOs in TracebackGroup... The solution _may_ have to be to use two parallel lists, one for the keys (exceptions) and one for the values (tracebacks). And a dict mapping the exception IDs to values. 

Actually I think now that EGs and TGs are immutable maybe the solution is to do the above and not worry about weakrefs.  The lifetime of the TG is identical to the lifetime of the EG, which holds references to the contained exceptions. So we never need an exception to GC itself out of a TG.

**Comment by gvanrossum** [comment]
Hm, I have to think about that more, with pen and paper.

I did find this discussion about weak key ID dicts, https://mail.python.org/pipermail/python-ideas/2014-December/030573.html (read the whole thread).

File: Lib/asyncio/taskgroup.py (hunk 3 above)
Comment (gvanrossum) [reply Lib/asyncio/taskgroup.py]: Which makes me think that taskgroup.py doesn’t belong in this PR.

File: Lib/exception_group.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+import sys
+
+
+class TracebackGroup:
+    def __init__(self, excs):
+        self.tb_next_map = {} # exception id to tb
+        for e in excs:
+            if isinstance(e, ExceptionGroup):
+                for e_ in e.excs:
+                    self.tb_next_map[id(e_)] = e_.__traceback__
```
Comment (gvanrossum) [inline Lib/exception_group.py]: What if ‘e_’ is an EG? Shouldn’t this be recursive? (Maybe you meant ‘for e_ in e’?)

File: Lib/exception_group.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+            else:
+                if e.__traceback__:
+                    self.tb_next_map[id(e)] = e.__traceback__.tb_next
```
Comment (gvanrossum) [inline Lib/exception_group.py]: The extra step (.tb_next) surprised me — maybe explain in a comment?

File: Lib/exception_group.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+            if isinstance(e, ExceptionGroup): # recurse
+                e_match, e_rest = e.project(condition)
+                if e_match:
+                    match.append(e_match)
+                if e_rest:
+                    rest.append(e_rest)
+            else:
+                if condition(e):
+                    match.append(e)
+                    e_match, e_rest = e, None
```
Comment (gvanrossum) [inline Lib/exception_group.py]: Looks like an unfinished thought or a dead end.

File: Lib/exception_group.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+        keep: List[BaseException]
+        """
+        match, _ = self.project(lambda e: e in keep)
+        return match
+
+    def extract_traceback(self, exc):
+        """ returns the traceback of a single exception
+
+        If exc is in this exception group, return its
+        traceback as a list of frames. Otherwise, return None.
```
Comment (gvanrossum) [inline Lib/exception_group.py]: Should probably also return the other info in the tb object (tb_lineno, tb_lasti). Possibly reuse one of the named tuple types from traceback.py?

File: Lib/exception_group.py [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+                tb = tb.tb_next
+            if isinstance(e, ExceptionGroup):
+                assert len(e.excs) == 1 and exc in e
+                e = e.excs[0]
+            else:
+                assert e is exc
+                e = None
+        return result
+
+    def push_frame(self, frame):
```
Comment (gvanrossum) [inline Lib/exception_group.py]: This is only used in tests?

File: Lib/exception_group.py [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+        if isinstance(exc, ExceptionGroup):
+            tbg = exc.__traceback_group__
+            assert isinstance(tbg, TracebackGroup)
+            indent += 4
+            for e in exc.excs:
+                t = tbg.tb_next_map[id(e)]
+                output.append('---------------------------------------')
+                output.extend(ExceptionGroup.render(e, t, indent))
+        for l in output:
+            print(l)
```
Comment (gvanrossum) [inline Lib/exception_group.py]: Debug print?

File: Lib/exception_group.py [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+                e = None
+        return result
+
+    def push_frame(self, frame):
+        import types
+        self.__traceback__ = types.TracebackType(
+            self.__traceback__, frame, 0, 0)
+
+    @staticmethod
+    def render(exc, tb=None, indent=0):
```
Comment (gvanrossum) [inline Lib/exception_group.py]: Eventually this needs to be integrated into traceback.py, or at least fit into the same framework.

And I guess we’ll need to teach the C code for printing tracebacks about EGs.

File: Lib/exception_group.py [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+            else:
+                if e.__traceback__:
+                    self.tb_next_map[id(e)] = e.__traceback__.tb_next
+                else:
+                    self.tb_next_map[id(e)] = None
+
+class ExceptionGroup(BaseException):
+
+    def __init__(self, excs, *, tb=None):
+        """ Construct a new ExceptionGroup
```
Comment (gvanrossum) [inline Lib/exception_group.py:25]: What would you think of adding a ‘message’ parameter here? Yury brought this up in the discussion.

File: Lib/exception_group.py [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+    def __iter__(self):
+        ''' iterate over the individual exceptions (flattens the tree) '''
+        for e in self.excs:
+            if isinstance(e, ExceptionGroup):
+                for e_ in e:
+                    yield e_
+            else:
+                yield e
+
+    def __len__(self):
```
Comment (gvanrossum) [inline Lib/exception_group.py]: I’m not a fan of a len() that is itself O(n). All we really need is to tell empty from non-empty, right? Let’s make that a normal method, or `__bool__`. We’re not trying to implement a `Sequence`.

File: Lib/exception_group.py [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+    def __exit__(self, etype, exc, tb):
+        if exc is not None and isinstance(exc, ExceptionGroup):
+            match, rest = exc.split(self.types)
+
+            if not match:
+                # Let the interpreter reraise the exception
+                return False
+
+            handler_excs = self.handler(match)
+            if handler_excs == match:
```
Comment (gvanrossum) [inline Lib/exception_group.py]: Um, EG doesn’t define `__eq__` so this uses `is`. Maybe that’s what we want, then write it that way?

File: Lib/exception_group.py [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+            else:
+                # to_keep: EG subgroup of exc with only those to reraise
+                # (either not matched or reraised by handler)
+                to_keep = exc.subgroup(
+                    list(rest) + [e for e in handler_excs if e in match])
+                # to_add: new exceptions raised by handler
+                to_add = handler_excs.subgroup(
+                    [e for e in handler_excs if e not in match])
+                if to_add:
+                    to_raise = ExceptionGroup([to_keep, to_add])
```
Comment (gvanrossum) [inline Lib/exception_group.py:213]: This introduces a spurious extra nesting level. It’s why I was asking for a merge operator.

**Review by gvanrossum — COMMENTED** [review]
Sorry, I can’t find the backtick on my iPad’s keyboard.

File: Lib/exception_group.py (hunk 12 above)
Comment (iritkatriel) [reply Lib/exception_group.py]: Yes, that line is not needed, I'll remove.

File: Lib/exception_group.py (hunk 14 above)
Comment (iritkatriel) [reply Lib/exception_group.py]: Not used at all. Removed.

File: Lib/exception_group.py (hunk 15 above)
Comment (iritkatriel) [reply Lib/exception_group.py]: This is the "render" function so it prints.   I want to extract the construction part out so it can be used in tests.

File: Lib/exception_group.py (hunk 20 above)
Comment (iritkatriel) [reply Lib/exception_group.py:213]: This is the case where the handler returned a mixture of old exceptions that we should reraise plus new exceptions that are unrelated to the exception group we're handling. It's the only case in ExceptionGroupCatcher where we create a new nesting level.

We have to merge the old and new into a new EG, because they don't have the same `__traceback__` (and `__context__`/`__cause__`, which I haven't handled yet at all - they also need to be copied by `project()`).   

The new exceptions either have no traceback, if they were just created, or they may have a different one one if they were raised and caught somewhere in the handler.

File: Lib/asyncio/taskgroup.py (hunk 3 above)
Comment (iritkatriel) [reply Lib/asyncio/taskgroup.py]: You're right, I copied it over from the previous branch but will remove.

File: Lib/exception_group.py (hunk 10 above)
Comment (iritkatriel) [reply Lib/exception_group.py]: Yes this looks wrong. I need more tests.

File: Lib/exception_group.py (hunk 15 above)
Comment (gvanrossum) [reply Lib/exception_group.py]: I guess that confused me -- it painstakingly constructs a list of lines and returns it, *and* also prints it. :-)

File: Lib/exception_group.py (hunk 20 above)
Comment (gvanrossum) [reply Lib/exception_group.py:213]: Hm, that's a tough one. Suppose you have:
```
def f():
    try:
        ...something that raises EG([E1, E2]) or perhaps EG([E2])...
    except *E1:
        ...handler code...
```
Then what bubbles out of `f()` is `EG([E2])`:

- if the try block raises `EG([E1, E2])` then what bubbles out is the remainder, `EG([E2])`
- if the try block raises `EG([E2])` then what bubbles out is exactly that

But if the handler raises `E3`, what bubbles out is wrapped in an extra `EG`, and `E3` is wrapped in *another* extra `EG`:
```
EG(EG([E2]), EG([E3]))
```
Note that the latter is apparently the responsibility of the handler -- you blatantly assume that `handler_excs` has a `subgroup()` method, i.e. that it is an `EG`.

I had somehow hoped that we'd get this instead:
```
EG([E2, E3])
```
But I see we can't do that because the tracebacks are different. I guess it's just something users who are manually unpacking the `EG` nesting structure will have to be aware of -- users using `except *` will be oblivious of this detail.

I'm a little sad about this because I had hoped that the `EG` nesting structure would reflect the TaskGroup nesting structure exactly, but clearly this cannot always be preserved. Oh well.

File: Lib/exception_group.py (hunk 17 above)
Comment (gvanrossum) [reply Lib/exception_group.py:25]: Hm, I had assumed the signature would have to be
```
def __init__(self, msg, excs, *, tb=None):
```
so that `msg` ends up in `self.args[0]`, where code dealing with exceptions in general expects a message.

File: Lib/exception_group.py (hunk 20 above)
Comment (iritkatriel) [reply Lib/exception_group.py:213]: Exactly. 

I was going to suggest that we allow the handler to return a list of exceptions, and then the result can be `EG(EG([E2]), E3))`.  We need to allow both exception list and EG, in case the handler actually did catch an EG with a traceback and so on.

Comment (gvanrossum) [reply Lib/exception_group.py:213]: Perhaps the handler can just raise exceptions instead of returning them?

Comment (iritkatriel) [reply Lib/exception_group.py:213]: Then the handler appears in the traceback, and I think that might make sense for new exceptions but not for ones it reraises.  Maybe we should allow both, because raising is very natural for new exceptions.

A use case that Nathaniel and Yuri were talking about was filtering OSErrors - the handler takes all OSErrors, checks their error code, handles one of them an tells us to reraise the rest.

Comment (gvanrossum) [reply Lib/exception_group.py:213]: Hm, but this thinking is geared towards the interim situation where users have to write a handler function. In the bright future we'll do away with handlers and they just use `except *E`. If that raises an EG it will appear in the traceback.

Realistically users will write broken handler functions that sometimes raise anyway, so we'll have to catch exceptions from the handler. A handler that truly re-raises can do so using bare `raise` -- IMO for everything else it's fine to have the `raise` statement appear in the traceback.

File: Lib/exception_group.py [hunk 21]
```diff
@@ -8,8 +8,11 @@ def __init__(self, excs):
         for e in excs:
             if isinstance(e, ExceptionGroup):
                 for e_ in e.excs:
-                    # TODO: what if e_ is an EG? This looks wrong.
-                    self.tb_next_map[id(e_)] = e_.__traceback__
+                    if isinstance(e_, ExceptionGroup):
+                        for k in e_:
+                            self.tb_next_map[id(k)] = e_.__traceback__
```
Comment (gvanrossum) [inline Lib/exception_group.py]: Sorry, but now my question is "what if k is an EG?" -- I think you need a recursive solution.

File: Lib/exception_group.py (hunk 20 above)
Comment (iritkatriel) [reply Lib/exception_group.py:213]: So the handler has some return value that means "naked raise", and it raises in all other cases?

**Review by gvanrossum — COMMENTED** [review]
> So the handler has some return value that means "naked raise", and it raises in all other cases?

Oh, that's a nice API -- the handler can choose whether to appear in the traceback or not.

File: Lib/exception_group.py (hunk 21 above)
Comment (iritkatriel) [reply Lib/exception_group.py]: `for k in e_:` does the recursion - it iterates over individual exceptions.

**Comment by iritkatriel** [comment]
> > So the handler has some return value that means "naked raise", and it raises in all other cases?
> 
> Oh, that's a nice API -- the handler can choose whether to appear in the traceback or not.

Not sure we're talking about the same thing. In my version it will be in the traceback unless it returns "naked raise", which makes the catcher return False (as under the current `if handler_excs is match:` case).

File: Lib/exception_group.py (hunk 21 above)
Comment (gvanrossum) [reply Lib/exception_group.py]: D'oh!

Comment (iritkatriel) [reply Lib/exception_group.py]: Oh, also - I removed the tb_next that you commented on from here. I think we might want to do something in render to avoid duplicates, but let's keep full info here. 

I think what happens is that the frame where an EG is created ends up on both sides of the seam (in `__traceback__` and `__traceback_group__`).

**Comment by gvanrossum** [comment]
> > > So the handler has some return value that means "naked raise", and it raises in all other cases?
> > 
> > 
> > Oh, that's a nice API -- the handler can choose whether to appear in the traceback or not.
> 
> Not sure we're talking about the same thing. In my version it will be in the traceback unless it returns "naked raise", which makes the catcher return False (as under the current `if handler_excs is match:` case).

Hm, if the handler creates a fresh exception and returns that, IIUC its frame won't be in the traceback; only `raise e` updates `e.__traceback__`. I guess when our `__exit__` code decides to raise a different exception than what was passed in (because something was handled or because something was added) it must use `raise <something>` and then *it* will appear in the traceback.

Basically, don't pay attention to me, I am just talking to myself here. :-)

**Comment by iritkatriel** [comment]
> Basically, don't pay attention to me, I am just talking to myself here. :-)

I'm talking to myself too, it's called 'conversation'. :)

**Comment by iritkatriel** [comment]
At the end of `ExceptionGroupCatcher.__exit__` there's a trick I copied from trio where it patches up `__context__` in the finally block. Can we mess with the traceback there as well, ie pop a frame from it?

**Comment by gvanrossum** [comment]
> At the end of `ExceptionGroupCatcher.__exit__` there's a trick I copied from trio where it patches up `__context__` in the finally block. Can we mess with the traceback there as well, ie pop a frame from it?

I don't see why not. That's a clever trick, doing this in the `finally` after you've already raised it.

**Comment by iritkatriel** [comment]
The trick doesn't seem to work for traceback.

```
import sys
import traceback

def f():
    raise ValueError()

def g():
    try:
        f()
    finally:
        _, value, _ = sys.exc_info()
        value.__traceback__ = None

try:
    g()
except ValueError as e:
    traceback.print_tb(e.__traceback__)
```

Output (I expected f() to not be there):

```
Running Release|Win32 interpreter...
  File "C:\Users\User\src\cpython\x.py", line 16, in <module>
    g()
  File "C:\Users\User\src\cpython\x.py", line 9, in g
    f()
  File "C:\Users\User\src\cpython\x.py", line 5, in f
    raise ValueError()
```

**Comment by gvanrossum** [comment]
> The trick doesn't seem to work for traceback.

Yeah, I discovered the same thing. Not sure why, possibly it has to do with the separation of exception type, value and traceback (maybe the tb is saved on the thread state while the finally block is running and restored from there, overwriting `__traceback__`).

File: Lib/exception_group.py [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+
+    def subgroup(self, keep):
+        """ Split an ExceptionGroup to extract only exceptions in keep
+
+        keep: List[BaseException]
+        """
+        match, _ = self.project(lambda e: e in keep)
+        return match
+
+    def extract_traceback(self, exc):
```
Comment (iritkatriel) [inline Lib/exception_group.py]: Not sure we need this as part of the API. It's used in the tests and can move there.

**Comment by gvanrossum** [comment]
Hey all, I've been distracted away from this. I'd like to make sure we don't lose momentum -- I don't want this to just sit here and get stale until the next sprint. So... How are things going here? What should our next step be? How close is this to the proposal Yury wrote up immediately after the sprint?

**Comment by iritkatriel** [comment]
I think what we have here implements stage1 of the plan in https://github.com/python/exceptiongroups/issues/9.
It follows the approach of trio, with the catcher context manager, so one thing we can perhaps do it to try and use this in trio to get an idea if it's making sense. 

The next phase of the plan is to integrate the rendering from here into traceback.py. I can try to do that on another branch, unless we think that's premature at this point.

**Comment by iritkatriel** [comment]
I think we can close this PR - we have a more advanced version of the python-only implementation here: https://github.com/iritkatriel/cpython/pull/3

**Comment by gvanrossum** [comment]
Agreed. This is really coming together -- thanks for plowing through!
